pkg/etcd/stream.go

```diff
@@ -42,16 +43,27 @@ func NewStream(logger kitlog.Logger, client etcdGetter, opt StreamOptions) (<-ch
 		defer cancel()
 		defer wg.Done()
 
-		logger.Log("event", "watch.start")
-		for resp := range client.Watch(opt.Ctx, "/", clientv3.WithPrefix()) {
+	Watch:
+
+		logger.Log("event", "watch_start")
+		for resp := range client.Watch(clientv3.WithRequireLeader(ctx), "/", clientv3.WithPrefix()) {
 			for _, event := range resp.Events {
+				if resp.Err() != nil {
+					logger.Log("error", resp.Err(), "msg", "received error from etcd watcher")
+				}
+
 				if includes(opt.Keys, string(event.Kv.Key)) {
 					out <- event.Kv
 				}
 			}
 		}
 
-		logger.Log("event", "watch.stop", "msg", "channel closed, stopping stream")
+		select {
+		case <-ctx.Done():
```

Comment:
I think this will only be reached if the watch channel is closed. Is that intentional? It seems like cancelling the context won't really do anything until the channel is closed - which could be an arbitrary time in the future.

Reply:
If the ctx is ever closed, the watch channel will also close. If the watch receives an unrecoverable error, it will also close.

So if we get to this select, we need to check if our context is done to see if we should end because we're being terminated or whether we ended because of an error.

Ah wait, I think I've just realised something:

> It seems like cancelling the context won't really do anything

Are you reading `ctx.Done()` as a method to cancel the context? As `ctx.Done()` returns a channel that receives a value if the context has completed, and doesn't actually cancel the context itself. Reading it like "cancel the context" would explain your comment.

Reply:
No, I knew that - what I missed was that we're passing the context into `Watch` (which should have been obvious, really).